File: crates/kuben-platform/src/materializer/render.rs
//! Rendering (ADR-032): the `Project`, `Environment` and `App` objects of
//! SQL rows alone.
//!
//! The objects keep the names, labels and shape the API has always given
//! them, so the existing controllers reconcile them unchanged. Projects and
//! environments are rendered the same way for a deployment run and for the
//! lifecycle operations that write them as soon as they exist. The release
//! decides an App's image: `spec.source` becomes
//! `<image repository>@<digest>`, whatever the configuration revision says
//! about the source.

use std::collections::BTreeMap;

use k8s_openapi::apimachinery::pkg::apis::meta::v1::OwnerReference;
use kube::api::ObjectMeta;
use kuben_core::ids::{EnvironmentId, OperationId, OrgId, ProjectId};
use kuben_crd::{
    App, AppSpec, DeletionPolicy, Environment, EnvironmentSpec, EnvironmentType, PreviewPolicy, Project,
    ProjectSpec, Protection, Quota, labels,
};
use kuben_store::repo::Materialization;
use serde_json::{Value, json};

use crate::controller::resources::{self, BuildError, RESTARTED_AT};

/// Field manager of every write the materializer makes.
pub const FIELD_MANAGER: &str = "kuben-materializer";

/// Annotations on materialized objects.
pub mod annotations {
    /// The target generation an App object was written for: the fence.
    pub const GENERATION: &str = "kuben.dev/generation";
    /// The operation that wrote the object last.
    pub const OPERATION: &str = "kuben.dev/operation";
    /// The target's lifecycle UID: a recreated target is another App.
    pub const LIFECYCLE_UID: &str = "kuben.dev/lifecycle-uid";
    /// The SQL id the object was rendered from: project, environment or target.
    pub const ID: &str = "kuben.dev/id";
    /// On an App handed over to the cluster's agent (M1.9): the target id.
    /// The App controller leaves such an App alone, so it never writes its
    /// workloads again while the App goes.
    pub const HANDOVER: &str = "kuben.dev/handover";
}

/// Grace period before a deleted production environment is purged, as the
/// API sets it.
const PRODUCTION_DELETION_GRACE: &str = "168h";
/// Process whose digest the App runs (the config revision contract).
const WEB: &str = "web";
/// Longest object name that is also a DNS label.
const MAX_NAME: usize = 63;

/// The objects of one run, in the order they are applied.
#[derive(Clone, Debug)]
pub struct Rendered {
    pub project: Project,
    pub environment: Environment,
    pub app: App,
}

/// What a `Project` object is rendered from.
#[derive(Clone, Copy, Debug)]
pub struct ProjectMaterial<'a> {
    pub org: OrgId,
    pub id: ProjectId,
    pub slug: &'a str,
    pub name: &'a str,
    pub description: Option<&'a str>,
}

/// What an `Environment` object is rendered from.
#[derive(Clone, Copy, Debug)]
pub struct EnvironmentMaterial<'a> {
    pub id: EnvironmentId,
    pub slug: &'a str,
    pub name: &'a str,
    /// `standard`, `production` or `preview`.
    pub env_type: &'a str,
    pub quota: Option<&'a Value>,
    /// The namespace of its placement, when it has one.
    pub namespace: Option<&'a str>,
}

impl<'a> ProjectMaterial<'a> {
    #[must_use]
    pub fn of(m: &'a Materialization) -> Self {
        Self {
            org: m.org,
            id: m.project,
            slug: &m.project_slug,
            name: &m.project_name,
            description: m.project_description.as_deref(),
        }
    }
}

impl<'a> EnvironmentMaterial<'a> {
    #[must_use]
    pub fn of(m: &'a Materialization) -> Self {
        Self {
            id: m.environment,
            slug: &m.environment_slug,
            name: &m.environment_name,
            env_type: &m.env_type,
            quota: m.quota.as_ref(),
            namespace: Some(&m.namespace),
        }
    }
}

/// Why an object cannot be rendered. The operation fails with
/// [`RenderError::code`].
#[derive(Clone, Debug, PartialEq, Eq, thiserror::Error)]
pub enum RenderError {
    #[error("release {0} has no image repository to pull its digest from")]
    NoImageRepository(String),
    #[error("release {0} has no `web` artifact")]
    NoArtifact(String),
    #[error("configuration revision {revision} is not an app spec: {reason}")]
    InvalidConfig { revision: u64, reason: String },
    #[error(transparent)]
    Invalid(#[from] BuildError),
    #[error("the placement namespace `{placement}` is not the environment's namespace `{expected}`")]
    Namespace { placement: String, expected: String },
    #[error("`{0}` is too long for a Kubernetes name")]
    NameTooLong(String),
    #[error("the environment quota is not valid: {0}")]
    InvalidQuota(String),
}

impl RenderError {
    /// Error code of the failed operation.
    #[must_use]
    pub fn code(&self) -> &'static str {
        match self {
            Self::NoImageRepository(_) => "NoImageRepository",
            Self::NoArtifact(_) => "NoArtifact",
            Self::InvalidConfig { .. } => "InvalidConfig",
            Self::Invalid(e) => e.reason(),
            Self::Namespace { .. } => "PlacementNamespace",
            Self::NameTooLong(_) => "NameTooLong",
            Self::InvalidQuota(_) => "InvalidQuota",
        }
    }
}

/// Kubernetes name of an environment: `<project>-<environment>`, as the API
/// names it.
#[must_use]
pub fn environment_name(project_slug: &str, environment_slug: &str) -> String {
    format!("{project_slug}-{environment_slug}")
}

/// Render the objects of the run `m`. Only one placement per environment
/// exists until M1.9, so the placement's namespace must be the one the
/// environment controller creates.
pub fn render(m: &Materialization) -> Result<Rendered, RenderError> {
    let project = ProjectMaterial::of(m);
    let environment = environment_object(&project, &EnvironmentMaterial::of(m), m.operation)?;
    if m.application_slug.len() > MAX_NAME {
        return Err(RenderError::NameTooLong(m.application_slug.clone()));
    }
    let environment_name = environment_name(&m.project_slug, &m.environment_slug);
    let mut app_labels = scope_labels(m.org, &m.project_slug);
    app_labels.insert(labels::ENVIRONMENT.to_owned(), environment_name);
    let mut app_annotations = BTreeMap::from([
        (annotations::GENERATION.to_owned(), m.generation.0.to_string()),
        (annotations::OPERATION.to_owned(), m.operation.to_string()),
        (annotations::LIFECYCLE_UID.to_owned(), m.lifecycle_uid.to_string()),
        (annotations::ID.to_owned(), m.target.to_string()),
    ]);
    if let Some(at) = m.restarted_at {
        // The builder copies it onto every pod template: a restart run
        // replaces the pods, a later run with the same stamp does not.
        app_annotations.insert(RESTARTED_AT.to_owned(), restart_stamp(at));
    }
    let app = App {
        metadata: ObjectMeta {
            name: Some(m.application_slug.clone()),
            namespace: Some(m.namespace.clone()),
            labels: Some(app_labels),
            annotations: Some(app_annotations),
            ..ObjectMeta::default()
        },
        spec: app_spec(m)?,
        status: None,
    };
    resources::validate(&app)?;
    Ok(Rendered {
        project: project_object(&project, m.operation),
        environment,
        app,
    })
}

/// A restart stamp as the API writes it on an App: RFC 3339, in seconds.
fn restart_stamp(ms: i64) -> String {
    k8s_openapi::jiff::Timestamp::from_millisecond(ms).map_or_else(
        |_| ms.to_string(),
        |t| t.strftime("%Y-%m-%dT%H:%M:%SZ").to_string(),
    )
}

/// The `Project` object of `p`, written by `operation`.
#[must_use]
pub fn project_object(p: &ProjectMaterial<'_>, operation: OperationId) -> Project {
    Project {
        metadata: ObjectMeta {
            name: Some(p.slug.to_owned()),
            labels: Some(managed_labels(p.org)),
            annotations: Some(written_by(operation, p.id.to_string())),
            ..ObjectMeta::default()
        },
        spec: ProjectSpec {
            display_name: p.name.to_owned(),
            description: p.description.map(str::to_owned),
            previews: PreviewPolicy::default(),
        },
        status: None,
    }
}

/// The `Environment` object of `e` in project `p`, written by `operation`.
pub fn environment_object(
    p: &ProjectMaterial<'_>,
    e: &EnvironmentMaterial<'_>,
    operation: OperationId,
) -> Result<Environment, RenderError> {
    let name = environment_name(p.slug, e.slug);
    let namespace = resources::namespace_name(&name);
    if namespace.len() > MAX_NAME {
        return Err(RenderError::NameTooLong(namespace));
    }
    if let Some(placement) = e.namespace
        && placement != namespace
    {
        return Err(RenderError::Namespace {
            placement: placement.to_owned(),
            expected: namespace,
        });
    }
    let (type_, protection) = match e.env_type {
        "production" => (
            EnvironmentType::Production,
            Some(Protection {
                require_approvals: 0,
                deletion_grace: PRODUCTION_DELETION_GRACE.into(),
            }),
        ),
        "preview" => (EnvironmentType::Preview, None),
        _ => (EnvironmentType::Standard, None),
    };
    let quota = e
        .quota
        .map(|q| serde_json::from_value::<Quota>(q.clone()))
        .transpose()
        .map_err(|err| RenderError::InvalidQuota(err.to_string()))?;
    Ok(Environment {
        metadata: ObjectMeta {
            name: Some(name),
            labels: Some(scope_labels(p.org, p.slug)),
            annotations: Some(written_by(operation, e.id.to_string())),
            ..ObjectMeta::default()
        },
        spec: EnvironmentSpec {
            project: p.slug.to_owned(),
            type_,
            deletion_policy: DeletionPolicy::Delete,
            protection,
            quota,
            ttl: None,
        },
        status: None,
    })
}

/// Make `environment` owned by `project`, as the API creates environments:
/// deleting the project deletes them. False while the project has no UID.
#[must_use]
pub fn set_owner(environment: &mut Environment, project: &Project) -> bool {
    let (Some(name), Some(uid)) = (&project.metadata.name, &project.metadata.uid) else {
        return false;
    };
    environment.metadata.owner_references = Some(vec![OwnerReference {
        api_version: "kuben.dev/v1alpha1".into(),
        kind: "Project".into(),
        name: name.clone(),
        uid: uid.clone(),
        ..OwnerReference::default()
    }]);
    true
}

fn managed_labels(org: OrgId) -> BTreeMap<String, String> {
    BTreeMap::from([
        (labels::MANAGED_BY.to_owned(), labels::MANAGER.to_owned()),
        (labels::ORG.to_owned(), org.to_string()),
    ])
}

fn scope_labels(org: OrgId, project_slug: &str) -> BTreeMap<String, String> {
    let mut l = managed_labels(org);
    l.insert(labels::PROJECT.to_owned(), project_slug.to_owned());
    l
}

fn written_by(operation: OperationId, id: String) -> BTreeMap<String, String> {
    BTreeMap::from([
        (annotations::OPERATION.to_owned(), operation.to_string()),
        (annotations::ID.to_owned(), id),
    ])
}

/// The configuration revision with the release's image as its source.
fn app_spec(m: &Materialization) -> Result<AppSpec, RenderError> {
    let digest = m
        .artifacts
        .get(WEB)
        .or_else(|| {
            let mut all = m.artifacts.values();
            match (all.next(), all.next()) {
                (Some(only), None) => Some(only),
                _ => None,
            }
        })
        .ok_or_else(|| RenderError::NoArtifact(m.release.to_string()))?;
    let repository = m
        .image_repository
        .as_deref()
        .filter(|r| !r.is_empty())
        .ok_or_else(|| RenderError::NoImageRepository(m.release.to_string()))?;
    let invalid = |reason: String| RenderError::InvalidConfig {
        revision: m.config_revision_number,
        reason,
    };
    let mut config = m.config.clone();
    let object = config
        .as_object_mut()
        .ok_or_else(|| invalid("not a JSON object".into()))?;
    object.insert(
        "source".into(),
        json!({ "image": format!("{repository}@{}", digest.as_str()) }),
    );
    let mut spec: AppSpec = serde_json::from_value(config).map_err(|e| invalid(e.to_string()))?;
    // A managed secret is read from the immutable object of the revision the
    // run is bound to; other names stay the cluster's own Secrets.
    for reference in spec.env.iter_mut().filter_map(|e| e.from_secret.as_mut()) {
        if let Some(bound) = m.secrets.iter().find(|b| b.name == reference.name) {
            reference.name = crate::secrets::object_name(&bound.name, bound.revision);
        }
    }
    Ok(spec)
}

#[cfg(test)]
mod tests {
    use kuben_core::{
        ids::{ApplicationId, ConfigRevisionId, DeploymentRunId, ReleaseId, TargetId},
        ops::{Generation, RunPhase},
    };
    use kuben_crd::Source;

    use super::*;

    const DIGEST: &str = "sha256:0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef";

    fn sample() -> Materialization {
        Materialization {
            render_plan: None,
            restarted_at: None,
            approval_expires_at: None,
            secrets: Vec::new(),
            cluster: kuben_core::ids::ClusterId::new(),
            delivery: kuben_store::repo::Delivery::Controller,
            org: OrgId::new(),
            run: DeploymentRunId::new(),
            operation: OperationId::new(),
            phase: RunPhase::Planned,
            generation: Generation(3),
            lifecycle_uid: uuid::Uuid::now_v7(),
            project: ProjectId::new(),
            project_slug: "shop".into(),
            project_name: "Shop".into(),
            project_description: Some("The online shop".into()),
            environment: EnvironmentId::new(),
            environment_slug: "production".into(),
            environment_name: "Production".into(),
            protected: true,
            env_type: "production".into(),
            quota: None,
            namespace: "kb-shop-production".into(),
            application: ApplicationId::new(),
            application_slug: "web".into(),
            application_name: "Web".into(),
            target: TargetId::new(),
            desired_generation: Generation(3),
            deleting: false,
            release: ReleaseId::new(),
            artifacts: BTreeMap::from([("web".to_owned(), DIGEST.parse().expect("digest"))]),
            image_repository: Some("ghcr.io/acme/web".into()),
            config_revision: ConfigRevisionId::new(),
            config_revision_number: 2,
            config: json!({
                "runtime": { "processes": { "web": { "port": 8080 } } },
                "env": [{ "name": "LOG_LEVEL", "value": "info" }],
            }),
        }
    }

    fn with_config(config: Value) -> Materialization {
        Materialization { config, ..sample() }
    }

    #[test]
    fn bound_secrets_are_read_from_their_revision_objects() {
        let m = Materialization {
            secrets: vec![kuben_store::repo::SecretBinding {
                name: "db".into(),
                secret: uuid::Uuid::now_v7(),
                revision: 4,
            }],
            ..with_config(json!({
                "runtime": { "processes": { "web": { "port": 8080 } } },
                "env": [
                    { "name": "DATABASE_URL", "fromSecret": { "name": "db", "key": "url" } },
                    { "name": "TOKEN", "fromSecret": { "name": "legacy", "key": "token" } },
                ],
            }))
        };
        let app = render(&m).expect("render").app;
        let names: Vec<_> = app
            .spec
            .env
            .iter()
            .filter_map(|e| e.from_secret.as_ref().map(|r| (r.name.as_str(), r.key.as_str())))
            .collect();
        assert_eq!(names, [("db.r4", "url"), ("legacy", "token")]);
    }

    #[test]
    fn a_restart_run_stamps_the_app() {
        let stamp = |m: &Materialization| {
            render(m)
                .expect("render")
                .app
                .metadata
                .annotations
                .and_then(|a| a.get(RESTARTED_AT).cloned())
        };
        assert_eq!(stamp(&sample()), None);
        let restarted = Materialization {
            restarted_at: Some(1_757_937_600_000),
            ..sample()
        };
        assert_eq!(stamp(&restarted).as_deref(), Some("2025-09-15T12:00:00Z"));
    }

    #[test]
    fn renders_the_objects_the_controllers_expect() {
        let m = sample();
        let r = render(&m).expect("render");

        assert_eq!(r.project.metadata.name.as_deref(), Some("shop"));
        assert_eq!(r.project.spec.display_name, "Shop");
        assert_eq!(r.project.spec.description.as_deref(), Some("The online shop"));
        assert_eq!(r.environment.metadata.name.as_deref(), Some("shop-production"));
        assert_eq!(r.environment.spec.project, "shop");
        assert_eq!(r.environment.spec.type_, EnvironmentType::Production);
        assert!(r.environment.spec.protection.is_some());
        assert_eq!(r.app.metadata.name.as_deref(), Some("web"));
        assert_eq!(r.app.metadata.namespace.as_deref(), Some("kb-shop-production"));

        let app_labels = r.app.metadata.labels.as_ref().expect("labels");
        assert_eq!(app_labels[labels::MANAGED_BY], labels::MANAGER);
        assert_eq!(app_labels[labels::ORG], m.org.to_string());
        assert_eq!(app_labels[labels::PROJECT], "shop");
        assert_eq!(app_labels[labels::ENVIRONMENT], "shop-production");
        let notes = r.app.metadata.annotations.as_ref().expect("annotations");
        assert_eq!(notes[annotations::GENERATION], "3");
        assert_eq!(notes[annotations::OPERATION], m.operation.to_string());
        assert_eq!(notes[annotations::LIFECYCLE_UID], m.lifecycle_uid.to_string());
        assert_eq!(notes[annotations::ID], m.target.to_string());
        for object in [&r.project.metadata, &r.environment.metadata] {
            let notes = object.annotations.as_ref().expect("annotations");
            assert_eq!(notes[annotations::OPERATION], m.operation.to_string());
            assert!(
                !notes.contains_key(annotations::GENERATION),
                "generations belong to targets"
            );
        }

        assert_eq!(
            r.app.spec.source,
            Source::from_image(format!("ghcr.io/acme/web@{DIGEST}"))
        );
        assert_eq!(r.app.spec.runtime.processes["web"].port, Some(8080));
        assert_eq!(r.app.spec.env[0].value.as_deref(), Some("info"));
    }

    #[test]
    fn the_release_decides_the_image() {
        let m = with_config(json!({
            "source": { "git": { "repo": "https://github.com/acme/web" } },
            "runtime": { "processes": { "worker": {} } },
        }));
        let r = render(&m).expect("render");
        assert_eq!(
            r.app.spec.source,
            Source::from_image(format!("ghcr.io/acme/web@{DIGEST}"))
        );

        let only = Materialization {
            artifacts: BTreeMap::from([("api".to_owned(), DIGEST.parse().expect("digest"))]),
            ..sample()
        };
        assert!(
            render(&only).is_ok(),
            "a release with one artifact names its image"
        );
    }

    #[test]
    fn environments_carry_their_type_and_quota() {
        let m = sample();
        let project = ProjectMaterial::of(&m);
        let operation = OperationId::new();
        let preview = EnvironmentMaterial {
            env_type: "preview",
            namespace: None,
            quota: Some(&json!({ "cpu": "4", "memory": "4Gi", "pods": 30 })),
            ..EnvironmentMaterial::of(&m)
        };
        let env = environment_object(&project, &preview, operation).expect("render");
        assert_eq!(env.spec.type_, EnvironmentType::Preview);
        assert!(env.spec.protection.is_none());
        let quota = env.spec.quota.expect("quota");
        assert_eq!(
            (quota.cpu.as_deref(), quota.memory.as_deref(), quota.pods),
            (Some("4"), Some("4Gi"), Some(30))
        );
        assert_eq!(
            env.metadata.annotations.expect("annotations")[annotations::OPERATION],
            operation.to_string()
        );

        let standard = EnvironmentMaterial {
            env_type: "standard",
            ..EnvironmentMaterial::of(&m)
        };
        let env = environment_object(&project, &standard, operation).expect("render");
        assert_eq!(env.spec.type_, EnvironmentType::Standard);

        let broken = EnvironmentMaterial {
            quota: Some(&json!({ "pods": "many" })),
            ..EnvironmentMaterial::of(&m)
        };
        assert_eq!(
            environment_object(&project, &broken, operation)
                .expect_err("quota")
                .code(),
            "InvalidQuota"
        );
    }

    #[test]
    fn an_object_that_cannot_be_rendered_says_why() {
        let codes = [
            (
                Materialization {
                    image_repository: None,
                    ..sample()
                },
                "NoImageRepository",
            ),
            (
                Materialization {
                    artifacts: BTreeMap::new(),
                    ..sample()
                },
                "NoArtifact",
            ),
            (with_config(json!([])), "InvalidConfig"),
            (with_config(json!({ "env": [] })), "InvalidConfig"),
            (
                with_config(json!({ "runtime": { "processes": {} } })),
                "NoProcesses",
            ),
            (
                Materialization {
                    namespace: "elsewhere".into(),
                    ..sample()
                },
                "PlacementNamespace",
            ),
            (
                Materialization {
                    environment_slug: "x".repeat(60),
                    ..sample()
                },
                "NameTooLong",
            ),
        ];
        for (m, code) in codes {
            let err = render(&m).expect_err(code);
            assert_eq!(err.code(), code, "{err}");
        }
    }

    #[test]
    fn environments_are_owned_by_their_project() {
        let r = render(&sample()).expect("render");
        let mut environment = r.environment.clone();
        assert!(
            !set_owner(&mut environment, &r.project),
            "no UID before the project is applied"
        );
        let mut project = r.project;
        project.metadata.uid = Some("uid-1".into());
        assert!(set_owner(&mut environment, &project));
        let owners = environment.metadata.owner_references.expect("owner");
        assert_eq!(
            (owners[0].kind.as_str(), owners[0].uid.as_str()),
            ("Project", "uid-1")
        );
    }
}

File: go/hub/internal/platform/materializer/render.go
package materializer

// Rendering (ADR-032): the Project, Environment and App objects of SQL
// rows alone.
//
// The objects keep the names, labels and shape the API has always given
// them, so the existing controllers reconcile them unchanged. Projects and
// environments are rendered the same way for a deployment run and for the
// lifecycle operations that write them as soon as they exist. The release
// decides an App's image: `spec.source` becomes
// `<image repository>@<digest>`, whatever the configuration revision says
// about the source.

import (
	"encoding/json"
	"errors"
	"fmt"
	"maps"
	"slices"
	"strconv"
	"time"

	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/types"

	"github.com/Teamtem-dev/kuben/go/hub/internal/core/ids"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/opt"
	"github.com/Teamtem-dev/kuben/go/hub/internal/platform/render"
	"github.com/Teamtem-dev/kuben/go/hub/internal/store"
	"github.com/Teamtem-dev/kuben/go/kubenapi/v1alpha1"
)

const (
	// productionDeletionGrace is the grace period before a deleted
	// production environment is purged, as the API sets it.
	productionDeletionGrace = "168h"
	// webProcess is the process whose digest the App runs (the config
	// revision contract).
	webProcess = "web"
	// maxName is the longest object name that is also a DNS label.
	maxName = 63
)

// Rendered is the objects of one run, in the order they are applied.
type Rendered struct {
	Project     v1alpha1.Project
	Environment v1alpha1.Environment
	App         v1alpha1.App
}

// ProjectMaterial is what a Project object is rendered from.
type ProjectMaterial struct {
	Org         ids.OrgID
	ID          ids.ProjectID
	Slug        string
	Name        string
	Description opt.Val[string]
}

// EnvironmentMaterial is what an Environment object is rendered from.
type EnvironmentMaterial struct {
	ID   ids.EnvironmentID
	Slug string
	Name string
	// EnvType is `standard`, `production` or `preview`.
	EnvType string
	Quota   opt.Val[any]
	// Namespace is the namespace of its placement, when it has one.
	Namespace opt.Val[string]
}

// ProjectMaterialOf is the project material of the run m.
func ProjectMaterialOf(m *store.Materialization) ProjectMaterial {
	return ProjectMaterial{
		Org: m.Org, ID: m.Project, Slug: m.ProjectSlug, Name: m.ProjectName, Description: m.ProjectDescription,
	}
}

// EnvironmentMaterialOf is the environment material of the run m.
func EnvironmentMaterialOf(m *store.Materialization) EnvironmentMaterial {
	return EnvironmentMaterial{
		ID: m.Environment, Slug: m.EnvironmentSlug, Name: m.EnvironmentName, EnvType: m.EnvType,
		Quota: m.Quota, Namespace: opt.Some(m.Namespace),
	}
}

// RenderError is why an object cannot be rendered; the operation fails
// with its Code.
type RenderError struct {
	// Code is the error code of the failed operation: NoImageRepository,
	// NoArtifact, InvalidConfig, PlacementNamespace, NameTooLong,
	// InvalidQuota, or the reason of a spec the builder refuses.
	Code    string
	message string
}

func (e *RenderError) Error() string { return e.message }

func renderError(code, format string, args ...any) *RenderError {
	return &RenderError{Code: code, message: fmt.Sprintf(format, args...)}
}

// EnvironmentName is the Kubernetes name of an environment:
// `<project>-<environment>`, as the API names it.
func EnvironmentName(projectSlug, environmentSlug string) string {
	return projectSlug + "-" + environmentSlug
}

// Render renders the objects of the run m. Only one placement per
// environment exists until M1.9, so the placement's namespace must be the
// one the environment controller creates.
func Render(m *store.Materialization) (Rendered, *RenderError) {
	project := ProjectMaterialOf(m)
	environment, rerr := EnvironmentObject(project, EnvironmentMaterialOf(m), m.Operation)
	if rerr != nil {
		return Rendered{}, rerr
	}
	if len(m.ApplicationSlug) > maxName {
		return Rendered{}, renderError("NameTooLong", "`%s` is too long for a Kubernetes name", m.ApplicationSlug)
	}
	labels := scopeLabels(m.Org, m.ProjectSlug)
	labels[v1alpha1.LabelEnvironment] = EnvironmentName(m.ProjectSlug, m.EnvironmentSlug)
	annotations := map[string]string{
		v1alpha1.AnnotationGeneration:   strconv.FormatUint(uint64(m.Generation), 10),
		v1alpha1.AnnotationOperation:    m.Operation.String(),
		v1alpha1.AnnotationLifecycleUID: m.LifecycleUID.String(),
		v1alpha1.AnnotationID:           m.Target.String(),
	}
	if at, ok := m.RestartedAt.Get(); ok {
		// The builder copies it onto every pod template: a restart run
		// replaces the pods, a later run with the same stamp does not.
		annotations[render.RestartedAt] = restartStamp(at)
	}
	spec, rerr := appSpec(m)
	if rerr != nil {
		return Rendered{}, rerr
	}
	app := v1alpha1.App{
		ObjectMeta: metav1.ObjectMeta{
			Name: m.ApplicationSlug, Namespace: m.Namespace, Labels: labels, Annotations: annotations,
		},
		Spec: spec,
	}
	if err := render.Validate(&app); err != nil {
		var build *render.BuildError
		if errors.As(err, &build) && build != nil {
			return Rendered{}, &RenderError{Code: string(build.Reason), message: build.Error()}
		}
		return Rendered{}, renderError("InvalidConfig", "%v", err)
	}
	return Rendered{Project: ProjectObject(project, m.Operation), Environment: environment, App: app}, nil
}

// restartStamp is a restart stamp as the API writes it on an App: RFC 3339,
// in seconds.
func restartStamp(ms int64) string {
	t := time.UnixMilli(ms).UTC()
	if y := t.Year(); y < -9999 || y > 9999 {
		// Outside the range jiff can print: Rust wrote the number.
		return strconv.FormatInt(ms, 10)
	}
	return t.Format("2006-01-02T15:04:05Z")
}

// ProjectObject is the Project object of p, written by operation.
func ProjectObject(p ProjectMaterial, operation ids.OperationID) v1alpha1.Project {
	var description *string
	if d, ok := p.Description.Get(); ok {
		description = &d
	}
	return v1alpha1.Project{
		ObjectMeta: metav1.ObjectMeta{
			Name: p.Slug, Labels: managedLabels(p.Org), Annotations: writtenBy(operation, p.ID.String()),
		},
		Spec: v1alpha1.ProjectSpec{
			DisplayName: p.Name, Description: description, Previews: v1alpha1.DefaultPreviewPolicy(),
		},
	}
}

// EnvironmentObject is the Environment object of e in project p, written by
// operation.
func EnvironmentObject(p ProjectMaterial, e EnvironmentMaterial, operation ids.OperationID) (v1alpha1.Environment, *RenderError) {
	name := EnvironmentName(p.Slug, e.Slug)
	namespace := render.NamespaceName(name)
	if len(namespace) > maxName {
		return v1alpha1.Environment{}, renderError("NameTooLong", "`%s` is too long for a Kubernetes name", namespace)
	}
	if placement, ok := e.Namespace.Get(); ok && placement != namespace {
		return v1alpha1.Environment{}, renderError("PlacementNamespace",
			"the placement namespace `%s` is not the environment's namespace `%s`", placement, namespace)
	}
	spec := v1alpha1.EnvironmentSpec{Project: p.Slug, DeletionPolicy: v1alpha1.DeletionPolicyDelete}
	switch e.EnvType {
	case "production":
		spec.Type = v1alpha1.EnvironmentTypeProduction
		spec.Protection = &v1alpha1.Protection{DeletionGrace: productionDeletionGrace}
	case "preview":
		spec.Type = v1alpha1.EnvironmentTypePreview
	default:
		spec.Type = v1alpha1.EnvironmentTypeStandard
	}
	if q, ok := e.Quota.Get(); ok {
		quota, err := decodeQuota(q)
		if err != nil {
			return v1alpha1.Environment{}, renderError("InvalidQuota", "the environment quota is not valid: %v", err)
		}
		spec.Quota = quota
	}
	return v1alpha1.Environment{
		ObjectMeta: metav1.ObjectMeta{
			Name: name, Labels: scopeLabels(p.Org, p.Slug), Annotations: writtenBy(operation, e.ID.String()),
		},
		Spec: spec,
	}, nil
}

// decodeQuota reads a stored quota as serde read a `Quota`: a JSON null is
// not one.
func decodeQuota(v any) (*v1alpha1.Quota, error) {
	if v == nil {
		return nil, errors.New("invalid type: null, expected struct Quota")
	}
	data, err := json.Marshal(v)
	if err != nil {
		return nil, err //nolint:wrapcheck // said as InvalidQuota
	}
	var quota v1alpha1.Quota
	if err := json.Unmarshal(data, &quota); err != nil {
		return nil, err //nolint:wrapcheck // said as InvalidQuota
	}
	return &quota, nil
}

// SetOwner makes environment owned by project, as the API creates
// environments: deleting the project deletes them. False while the
// project has no UID.
func SetOwner(environment *v1alpha1.Environment, project *v1alpha1.Project) bool {
	if project.Name == "" || project.UID == "" {
		return false
	}
	environment.OwnerReferences = []metav1.OwnerReference{{
		APIVersion: v1alpha1.SchemeGroupVersion.String(),
		Kind:       v1alpha1.ProjectKind,
		Name:       project.Name,
		UID:        types.UID(string(project.UID)),
	}}
	return true
}

func managedLabels(org ids.OrgID) map[string]string {
	return map[string]string{v1alpha1.LabelManagedBy: v1alpha1.LabelManagerValue, v1alpha1.LabelOrg: org.String()}
}

func scopeLabels(org ids.OrgID, projectSlug string) map[string]string {
	l := managedLabels(org)
	l[v1alpha1.LabelProject] = projectSlug
	return l
}

func writtenBy(operation ids.OperationID, id string) map[string]string {
	return map[string]string{v1alpha1.AnnotationOperation: operation.String(), v1alpha1.AnnotationID: id}
}

// SecretObjectName is the name of the immutable Secret holding revision
// of the managed secret name (crate::secrets::object_name).
func SecretObjectName(name string, revision uint64) string {
	return name + ".r" + strconv.FormatUint(revision, 10)
}

// appSpec is the configuration revision with the release's image as its
// source.
func appSpec(m *store.Materialization) (v1alpha1.AppSpec, *RenderError) {
	digest, ok := m.Artifacts[webProcess]
	if !ok && len(m.Artifacts) == 1 {
		for _, only := range m.Artifacts {
			digest, ok = only, true
		}
	}
	if !ok {
		return v1alpha1.AppSpec{}, renderError("NoArtifact", "release %s has no `web` artifact", m.Release)
	}
	repository := m.ImageRepository.Or("")
	if repository == "" {
		return v1alpha1.AppSpec{}, renderError("NoImageRepository",
			"release %s has no image repository to pull its digest from", m.Release)
	}
	invalid := func(reason string) *RenderError {
		return renderError("InvalidConfig", "configuration revision %d is not an app spec: %s",
			m.ConfigRevisionNumber, reason)
	}
	config, isObject := m.Config.(map[string]any)
	if !isObject {
		return v1alpha1.AppSpec{}, invalid("not a JSON object")
	}
	// A copy: the stored configuration is not changed.
	withImage := make(map[string]any, len(config)+1)
	maps.Copy(withImage, config)
	config = withImage
	config["source"] = map[string]any{"image": repository + "@" + digest.String()}
	data, err := json.Marshal(config)
	if err != nil {
		return v1alpha1.AppSpec{}, invalid(err.Error())
	}
	spec, err := v1alpha1.DecodeAppSpec(data)
	if err != nil {
		return v1alpha1.AppSpec{}, invalid(err.Error())
	}
	// A managed secret is read from the immutable object of the revision
	// the run is bound to; other names stay the cluster's own Secrets.
	// Images are pulled with the bound registry logins only.
	for i := range spec.Env {
		reference := spec.Env[i].FromSecret
		if reference == nil {
			continue
		}
		j := slices.IndexFunc(m.Secrets, func(b store.SecretBinding) bool {
			return b.Registry.IsNone() && b.Name == reference.Name
		})
		if j >= 0 {
			reference.Name = SecretObjectName(m.Secrets[j].Name, m.Secrets[j].Revision)
		}
	}
	spec.ImagePullSecrets = nil
	for _, b := range m.Secrets {
		if b.Registry.IsSome() {
			spec.ImagePullSecrets = append(spec.ImagePullSecrets, SecretObjectName(b.Name, b.Revision))
		}
	}
	return spec, nil
}
